zookeeper-server/src/main/java/org/apache/zookeeper/server/watch/WatchStats.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.watch;
+
+/**
+ * Statistics for multiple different watches on one node.
+ */
+public class WatchStats {
+    /**
+     * Stats that have no watchers attached.
+     *
+     * <p>This could be used as start point to compute new stats using {@link #addMode(WatcherMode)}.
+     */
+    public static final WatchStats NONE = new WatchStats(0);
+
+    static final WatchStats[] WATCH_STATS;
```

Comment:
Using static variables makes it vary hard to write tests that start multiple nodes.

Can we get rid of this static variable? Maybe you can create a wrapper class that contains the array

Reply:
Not sure whether I have got what you mean.

But both `WatchStats.flag` and `WatchStats.WATCH_STATS` are final. I tried to make `WatchStats` immutable and create all possible combinations upfront to avoid potential massive equal objects in case of large number of watches. Old implementation uses single `WatcherMode`. Not sure whether it is worth or not. But I think we have two options:
* Use fixed number(aka. eight) of `WatchStats` objects.
* Make `WatchStats` mutable and create one per session path.

I am ok with either.

Reply:
sorry, I misunderstood the meaning of this class.
it really looks like a enum.

I believe that the current design works.

but, why aren'y you using a enum ?

Reply:
* The enum names might look like a mess. `StandardAndPersistentAndPersistentRecursive` and so on.
* There is not direct access to some mebers from outside. For example `StandardAndPersistent` should only be generated from `WatchStats.addMode` or `WatchStats.removeMode` but not outside.

Reply:
Makes sense.
Thanks

Your proposal sounds like a good design